pkg/mountinfo/mountinfo.go

```diff
@@ -129,35 +132,40 @@ func GetMountInfo() ([]*MountInfo, error) {
 	return parseMountInfoFile(fMounts)
 }
 
-func isMountFS(mountInfos []*MountInfo, mntType string, mapRoot string) (bool, bool) {
-	var mapRootMountInfo *MountInfo
-
-	for _, mountInfo := range mountInfos {
-		if mountInfo.MountPoint == mapRoot {
-			mapRootMountInfo = mountInfo
-			break
-		}
+// IsMountFS returns two boolean values, checking
+// - whether the mapRoot is a mount point;
+// - whether that filesystem type is mntType.
+// Note that
+// - mapRoot should be an absolute and cleaned path;
+// - mapRoot should not be '/'.
+func IsMountFS(mntType int64, mapRoot string) (bool, bool, error) {
+	// Check if mapRoot is a directory.
+	var st, pst unix.Stat_t
+	err := unix.Lstat(mapRoot, &st)
+	if err != nil {
+		return false, false, err
 	}
-
-	if mapRootMountInfo == nil {
-		return false, false
+	if st.Mode&unix.S_IFDIR == 0 {
+		return false, false, nil
 	}
 
-	if mapRootMountInfo.FilesystemType == mntType {
-		return true, true
+	// Check if mapRoot is a mount point.
+	err = unix.Lstat(filepath.Dir(mapRoot), &pst)
+	if err != nil {
+		return false, false, err
+	}
+	if st.Dev == pst.Dev {
+		// parent dir has the same dev -- mapRoot is not a mount point
+		return false, false, nil
 	}
-	return true, false
-}
 
-// IsMountFS returns two boolean values:checks whether the current mapRoot:
-// - whether the current mapRoot has any mount
-// - whether that mount's filesystem is of type mntType
-func IsMountFS(mntType string, mapRoot string) (bool, bool, error) {
-	mountInfos, err := GetMountInfo()
+	// Check the fstype
+	fst := unix.Statfs_t{}
+	err = unix.Statfs(mapRoot, &fst)
 	if err != nil {
-		return false, false, err
+		return true, false, err
 	}
 
-	mounted, mntTypeInstance := isMountFS(mountInfos, mntType, mapRoot)
```

Comment:
The implementation of `isMountFS` could be removed as well, AFAIK. Also the tests in `mountinfo_test.go` for that function will probably need to be adjusted, as well as the comment for `TestIsMountInfo`.

Reply:
> The implementation of isMountFS could be removed as well, AFAIK

Not sure what you mean here, perhaps you overlooked it, but the `isMountFS` implementation is already removed by this PR, see 
https://github.com/cilium/cilium/pull/12249/commits/1d44c567ca7ae813000d260dbbde49ae22ab0f89#diff-d6be046d88ce3434aa0d4ad066831319L132

> Also the tests in mountinfo_test.go for that function will probably need to be adjusted

They were adjusted

> as well as the comment for TestIsMountInfo.

If you mean `TestIsMountFS` then the comment was already modified, see https://github.com/cilium/cilium/pull/12249/commits/1d44c567ca7ae813000d260dbbde49ae22ab0f89#diff-7fedb29c1d69fb43ab664233d60aa4feR604

Reply:
I'm sorry, probaly got confused by the way the side-by-side diff is formatted.